File: apps/android-frontend/android/app/src/main/java/au/com/codecapers/photosphere/jsengine/NetworkHost.java
package au.com.codecapers.photosphere.jsengine;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.net.NetworkInfo;
import android.os.Build;

//
// Reports what kind of connection this phone currently has, for the background sync.
//
// The "Only sync over Wi-Fi" setting refuses a sync on a cellular connection, and the loop that
// decides that runs while the app is off screen, so it cannot ask the WebView what the network is
// doing: there may be no WebView. It asks here instead, through the platform.
//
// The names returned are the four computeSyncAllowed understands. What to do about each of them is
// decided there, in TypeScript, and not here: this class answers one question and takes no view on
// what the answer means.
//
public final class NetworkHost {

    //
    // The connection is Wi-Fi, or something else that is not metered by a mobile carrier.
    //
    public static final String CONNECTION_WIFI = "wifi";

    //
    // The connection is a mobile carrier's, which is the one the Wi-Fi-only setting refuses.
    //
    public static final String CONNECTION_CELLULAR = "cellular";

    //
    // There is no connection at all.
    //
    public static final String CONNECTION_NONE = "none";

    //
    // There is a connection, but not one that can be identified as either of the two above.
    //
    // computeSyncAllowed permits this, deliberately: an unrecognised transport must not stop syncing
    // altogether, which is the same reason the desktop reports it.
    //
    public static final String CONNECTION_UNKNOWN = "unknown";

    //
    // The Android context the connectivity service is reached through.
    //
    private final Context context;

    //
    // Constructs the reporter over the given Android context.
    //
    public NetworkHost(Context context) {
        this.context = context;
    }

    //
    // Returns the current connection type as one of the four names above.
    //
    // A phone with no connectivity service at all reports "none" rather than "unknown", because the
    // one thing that is certain in that case is that nothing can be pushed anywhere.
    //
    public String connectionType() {
        ConnectivityManager connectivityManager =
            (ConnectivityManager)context.getSystemService(Context.CONNECTIVITY_SERVICE);
        if (connectivityManager == null) {
            return CONNECTION_NONE;
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            Network activeNetwork = connectivityManager.getActiveNetwork();
            if (activeNetwork == null) {
                return CONNECTION_NONE;
            }

            NetworkCapabilities capabilities = connectivityManager.getNetworkCapabilities(activeNetwork);
            if (capabilities == null) {
                return CONNECTION_NONE;
            }

            return classifyCapabilities(capabilities);
        }

        return legacyConnectionType(connectivityManager);
    }

    //
    // Names the transport a set of capabilities describes.
    //
    // Cellular is checked first. A phone sharing its mobile connection over a VPN, or bridging one
    // transport over another, can report more than one, and the setting exists to keep a backup off
    // the carrier's network: answering "wifi" for a connection that is also cellular is the one
    // mistake here that costs the user money.
    //
    private String classifyCapabilities(NetworkCapabilities capabilities) {
        if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR)) {
            return CONNECTION_CELLULAR;
        }

        if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)
            || capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)) {
            // Ethernet counts as Wi-Fi here rather than as its own name, because the only question
            // anything asks of this is whether the connection is the carrier's. An emulator on a
            // host's network reports ethernet, and calling that "unknown" would be true and useless.
            return CONNECTION_WIFI;
        }

        return CONNECTION_UNKNOWN;
    }

    //
    // The same answer on versions of Android with no NetworkCapabilities to ask.
    //
    // The app supports back to a version where getActiveNetworkInfo is the only way to ask, and it is
    // deprecated rather than absent, so this path is reached only there.
    //
    @SuppressWarnings("deprecation")
    private String legacyConnectionType(ConnectivityManager connectivityManager) {
        NetworkInfo activeNetworkInfo = connectivityManager.getActiveNetworkInfo();
        if (activeNetworkInfo == null || !activeNetworkInfo.isConnected()) {
            return CONNECTION_NONE;
        }

        int type = activeNetworkInfo.getType();
        if (type == ConnectivityManager.TYPE_MOBILE || type == ConnectivityManager.TYPE_MOBILE_DUN) {
            return CONNECTION_CELLULAR;
        }

        if (type == ConnectivityManager.TYPE_WIFI || type == ConnectivityManager.TYPE_ETHERNET) {
            return CONNECTION_WIFI;
        }

        return CONNECTION_UNKNOWN;
    }
}

File: apps/ios-frontend/ios/App/App/JsEngine/NetworkHost.swift
import Foundation
import Network

//
// Reports what kind of connection this phone currently has, for the background sync.
//
// The "Only sync over Wi-Fi" setting refuses a sync on a cellular connection, and the loop that
// decides that runs while the app is off screen, so it cannot ask the WebView what the network is
// doing: there may be no WebView. It asks here instead, through the platform.
//
// The names returned are the four computeSyncAllowed understands, and match NetworkHost.java exactly.
// What to do about each of them is decided there, in TypeScript, and not here.
//
final class NetworkHost {

    //
    // The connection is Wi-Fi, or something else that is not metered by a mobile carrier.
    //
    static let connectionWifi = "wifi"

    //
    // The connection is a mobile carrier's, which is the one the Wi-Fi-only setting refuses.
    //
    static let connectionCellular = "cellular"

    //
    // There is no connection at all.
    //
    static let connectionNone = "none"

    //
    // There is a connection, but not one that can be identified as either of the two above.
    //
    // computeSyncAllowed permits this, deliberately: an unrecognised transport must not stop syncing
    // altogether, which is the same reason the desktop reports it.
    //
    static let connectionUnknown = "unknown"

    //
    // The one monitor, shared by every engine that asks.
    //
    // NWPathMonitor reports the current path continuously once started, rather than answering a
    // question when asked, so it is started once and its latest path is read from. Starting one per
    // call would answer "unsatisfied" for the first moments of every call, which for a sync loop
    // means refusing to sync on a phone that is perfectly well connected.
    //
    private static let shared = NetworkHost()

    //
    // The monitor whose latest path is read.
    //
    private let monitor = NWPathMonitor()

    //
    // Guards the latest path below, which the monitor's queue writes and callers read.
    //
    private let stateLock = NSLock()

    //
    // The most recent path the monitor reported, or nil before it has reported one.
    //
    private var latestPath: NWPath?

    //
    // Starts the monitor. Private: everything goes through the shared instance.
    //
    private init() {
        monitor.pathUpdateHandler = { [weak self] path in
            guard let self = self else {
                return
            }
            self.stateLock.lock()
            self.latestPath = path
            self.stateLock.unlock()
        }
        monitor.start(queue: DispatchQueue(label: "au.com.codecapers.photosphere.network-host"))
    }

    //
    // Returns the current connection type as one of the four names above.
    //
    static func connectionType() -> String {
        return shared.currentConnectionType()
    }

    //
    // The current connection type, from the latest path the monitor reported.
    //
    // A path that has not arrived yet reports "unknown" rather than "none". The monitor delivers its
    // first update within moments of starting, and the difference matters: "none" refuses a sync,
    // and refusing one because the answer had not arrived yet would be a phone that syncs a few
    // seconds later than it should have, reported as a phone with no network.
    //
    func currentConnectionType() -> String {
        stateLock.lock()
        let path = latestPath
        stateLock.unlock()

        guard let path = path else {
            return NetworkHost.connectionUnknown
        }

        if path.status != .satisfied {
            return NetworkHost.connectionNone
        }

        // Cellular is checked first. A phone sharing its mobile connection over a VPN, or bridging
        // one transport over another, can report more than one, and the setting exists to keep a
        // backup off the carrier's network: answering "wifi" for a connection that is also cellular
        // is the one mistake here that costs the user money.
        if path.usesInterfaceType(.cellular) {
            return NetworkHost.connectionCellular
        }

        if path.usesInterfaceType(.wifi) || path.usesInterfaceType(.wiredEthernet) {
            // Ethernet counts as Wi-Fi here rather than as its own name, because the only question
            // anything asks of this is whether the connection is the carrier's. A simulator on a
            // host's network reports ethernet, and calling that "unknown" would be true and useless.
            return NetworkHost.connectionWifi
        }

        return NetworkHost.connectionUnknown
    }
}
